core/translate.c

```diff
@@ -387,16 +418,19 @@ translate_walk_good_state(dcontext_t *tdcontext, translate_walk_t *walk,
                           app_pc translate_pc)
 {
     return (!walk->unsupported_mangle ||
-            /* If we're at the instr AFTER the mangle region, we're ok */
+            /* If we're at the instr AFTER the mangle region, or at an instruction
+             * in the mangled region's EPILOGUE, we're ok. */
             (walk->in_mangle_region && translate_pc != walk->translation));
 }
 
 static void
-translate_walk_restore(dcontext_t *tdcontext, translate_walk_t *walk, app_pc translate_pc)
+translate_walk_restore(dcontext_t *tdcontext, translate_walk_t *walk, instr_t *inst,
+                       app_pc translate_pc)
```

Comment:
It feels dangerous to point at the next app instr w/o changing how the state restore is done: I suppose the outstanding-spills code should emulate the restores not yet executed, but what about the xsp_adjust?  That code assumes it should prepare for re-executing the instr.

Reply:
Re: the outstanding-spills code: this is assuming the restores exactly mirror the spills.  Maybe worth mentioning that as it may not always be the case for complex mangling.

Reply:
I've added a comment for the outstanding-spills code (that I didn't touch). I have added the !translate_walk_enters_mangling_epilogue to keep the function's current semantics. Since mangling epilogue now changes the translation, it adds a check to the exclude mangling epilogue from the outstanding_spills checks.

Reply:
In some sense, it seems like if we're going to xl8 to the *next* instr, we should be re-arranging the whole translate_walk_restore: we should be walking the rest of the epilogue and looking at the restores, instead of looking at what happened earlier.

Reply:
And thus ignore earlier xsp changes.